clang/lib/Lex/PPMacroExpansion.cpp

```diff
@@ -1721,11 +1721,13 @@ void Preprocessor::ExpandBuiltinMacro(Token &Tok) {
     Diag(Tok.getLocation(), diag::warn_pp_date_time);
     // MSVC, ICC, GCC, VisualAge C++ extension.  The generated string should be
     // of the form "Ddd Mmm dd hh::mm::ss yyyy", which is returned by asctime.
-    const char *Result;
+    const char *Result = "??? ??? ?? ??:??:?? ????";
+    char TimeString[64];
     if (getPreprocessorOpts().SourceDateEpoch) {
       time_t TT = *getPreprocessorOpts().SourceDateEpoch;
       std::tm *TM = std::gmtime(&TT);
-      Result = asctime(TM);
+      if (strftime(TimeString, sizeof TimeString, "%a %b %e %T %Y", TM) != 0)
```

Comment:
Hmm, so the last remaining problem is that of locale. I didn't realize that `asctime()` is locale independent, which means that switching to `strftime()` will change the output for some users. :-(

We could use `setlocale()` before the call to `strftime()` and then reset the locale after the call, but another approach would be to refactor `ComputeDATE_TIME` which does all the formatting manually and in a locale-independent way. Given what a pain locales are in practice, I sort of think the refactoring is actually a better approach. WDYT @cor3ntin?

Reply:
This is why i suggested the C++ approach here https://github.com/llvm/llvm-project/pull/99075#issuecomment-2233382915 

(wouldn't it be nice if the C standard provided overloads of everything that takes explicit local parameters to avoid having to do these sort of dances!)

Reply:
I found that using `%b` in `put_time` is also locale-dependent. Is it expected?

Reply:
SHOOT, yes, that is expected, same for `%a`.

> This is why i suggested the C++ approach here https://github.com/llvm/llvm-project/pull/99075#issuecomment-2233382915

We can't use that approach because we need to compile in C++17 mode: https://compiler-explorer.com/z/xosc68WMn

Okay, so how's this for horrible:
```
bool Failed = false;
if (const char *OldLocale = ::setlocale(LC_TIME, "C")) {
  if (strftime(...)) {
  } else {
    Failed = true;
  }
  ::setlocale(LC_TIME, OldLocale);
} else {
  Failed = true;
}

if (Failed) {
  // Emit question marks instead
}
```

Reply:
@AaronBallman the `std::print` thing was illistrative, this works in C++17 (well, 11)

```cpp
    std::stringstream buf;
    // this is the line that ensure we are not affected by locales 
    buf.imbue(std::locale("C"));
    std::time_t t = std::time(nullptr);
    std::tm tm = *std::localtime(&t);
    buf << std::put_time(&tm, "%c");
    std::string str = buf.str(); // do whatever with that string
```

Reply:
Ah, okay, thanks! Hmm, if we have to mess with locales either way, this seems reasonable to me.

Reply:
`asctime_r` might be the best replacement, but it seems unavailable on MSVC.
MSVC has `asctime_s`.